File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string s) {
    int answer = 0;
    
    char x = s[0];
    int same = 0;
    int diff = 0;
    
    for(int i = 0; i < s.size(); i++) {
        if(x == s[i]) {
            same++;
        } else {
            diff++;
        }
        
        // 두 횟수가 같은 경우 문자열 분리
        if(same == diff) {
            answer++;
            
            x = s[i + 1];
            same = 0;
            diff = 0;
        }
    }
    
    // 두 횟수가 다른 상태에서 더 이상 읽을 글자가 없으면 문자열 분리(+1)
    if(same != diff) {
        answer++;
    }
        
    return answer;
}

File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.java
class Solution {
    public int solution(String s) {
        int answer = 0;

        char c = s.charAt(0);
        int x = 1;
        int notX = 0;

        for(int i = 1; i < s.length(); i++) {
            if(x == notX) {
                answer++;
                c = s.charAt(i);
                x = 0;
                notX = 0;
            }

            if(c == s.charAt(i)) {
                x++;
            } else {
                notX++;
            }
        }

        // 반복문 종료되고 마지막 남은 문자열
        answer++;

        return answer;
    }
}